### DISCUSSION TITLE
libxml2 issue with Homebrew on Linux

### DISCUSSION THREAD

**Discussion by jcamiel** [Homebrew/discussions#2627] [Linux] [answered]
Hi,

I'm trying to build a Homebrew formula for [Hurl](https://hurl.dev).
Hurl uses Rust libxml crate for xpath evaluation, which is a wrapper on C libxml2.

On macOS, I don't have any problem and my formula is building.
On Linux, I've some problem building my binary.

My brew formula is:

```ruby
...
uses_from_macos "libxml2"

def install
    system "cargo", "install", *std_cargo_args(path: "packages/hurl")
end
```

If I try to build as it, I've a compilation error (raised by libxml/build.rs, a build script used by cargo):

```
thread 'main' panicked at 'Could not find libxml2.', /root/.cache/Homebrew/cargo_cache/registry/src/github.com-1ecc6299db9ec823/libxml-0.3.0/build.rs:25:5
...
HOMEBREW_CC: gcc-5
HOMEBREW_CXX: g++-5
MAKEFLAGS: -j2
CMAKE_PREFIX_PATH: /home/linuxbrew/.linuxbrew
PKG_CONFIG_PATH: /home/linuxbrew/.linuxbrew/opt/curl/lib/pkgconfig:/home/linuxbrew/.linuxbrew/opt/libxml2/lib/pkgconfig:/home/linuxbrew/.linuxbrew/opt/openssl@1.1/lib/pkgconfig:/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/pkgconfig
HOMEBREW_GIT: git
ACLOCAL_PATH: /home/linuxbrew/.linuxbrew/share/aclocal
PATH: /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super:/home/linuxbrew/.linuxbrew/opt/rust/bin:/home/linuxbrew/.linuxbrew/opt/curl/bin:/home/linuxbrew/.linuxbrew/opt/libxml2/bin:/home/linuxbrew/.linuxbrew/opt/openssl@1.1/bin:/home/linuxbrew/.linuxbrew/opt/python@3.9/bin:/home/linuxbrew/.linuxbrew/opt/binutils/bin:/usr/bin:/bin:/usr/sbin:/sbin:/home/linuxbrew/.linuxbrew/opt/gcc@5/bin
```

This build script use the Rust crate pkg_config to find libxml2 build metadata:

https://github.com/KWARC/rust-libxml/blob/40020e15b726c27dc4435e9d104a9799f4c77f46/build.rs#L29-L35

```rust
#[cfg(any(target_family="unix", target_os="macos"))]
mod pkg_config_dep {
  pub fn find() -> bool {
    if pkg_config::find_library("libxml-2.0").is_ok() {
      return true;
    }
    false
  }
}
```
I struggle to understand why this snippet doesn't seems to work: in the brew log, I can see `PKG_CONFIG_PATH: /home/linuxbrew/.linuxbrew/opt/curl/lib/pkgconfig:/home/linuxbrew/.linuxbrew/opt/libxml2/lib/pkgconfig:/home/linuxbrew/.linuxbrew/opt/openssl@1.1/lib/pkgconfig:/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/pkgconfig` with libxml pkgconfig. The line `pkg_config::find_library("libxml-2.0")` should return the right metadata to cargo.  

In the same configuration, if I build my app directly with cargo (`cargo build`) instead of building through the Homebrew formula, I can build and link to Homebrew's libxml2 without any issue.

I acknowledge I'm not an expert neither in Homebrew nor Rust, even in Linux stuff so I maybe miss something trivial.
I've tried a lot of things and I think I'm becoming mad, so sorry if this is not the right place for my issue

Thank you,

__edit__: I've added a kind of minimal setup to reproduce my problem:

1. A simple rust program that uses libxml crate (<https://github.com/jcamiel/testxml>):

src/main.rs:

```rust
use libxml::tree::Document;

fn main() {
    let doc_result = Document::new();
    assert!(doc_result.is_ok());
}
```

Cargo.toml
```toml
[package]
name = "testxml"
version = "0.1.0"
edition = "2021"

[dependencies]
libxml = "0.3.0"
```

2. A simple Brew formula (<https://github.com/jcamiel/homebrew-testxml>):

```ruby
class Testxml < Formula
  desc "Test xml"
  url "https://github.com/jcamiel/testxml/archive/refs/tags/0.0.1.tar.gz"
  sha256 "72c145617d9af341d332c330f164a987a11b6abe590522a8b675ccd8f00edb78"

  depends_on "rust" => :build

  on_linux do
      depends_on "libxml2" => :build
  end

  def install
      system "cargo", "install", *std_cargo_args(path: ".")
  end

end
```

3. Test this formula on macOS:

```
$ brew tap jcamiel/testxml
$ brew install testxml
$ testxml
```

4. Test this formula on Linux:

I use this Dockerfile to test a new Linux image:

```Dockerfile
FROM debian:bullseye

RUN apt-get update
RUN apt-get -y install curl \
    git \
    clang
RUN /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
ENV PATH="/home/linuxbrew/.linuxbrew/bin:${PATH}"
```

Then, install the test formula:

```
$ brew tap jcamiel/testxml
$ brew install testxml
...
==> cargo install
Last 15 lines from /root/.cache/Homebrew/Logs/testxml/01.cargo:
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR

  --- stderr
  thread 'main' panicked at 'Could not find libxml2.', /root/.cache/Homebrew/cargo_cache/registry/src/github.com-1ecc6299db9ec823/libxml-0.3.0/build.rs:25:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: failed to compile `testxml v0.1.0 (/tmp/testxml-20211212-10682-nobigf/testxml-0.0.1)`, intermediate artifacts can be found at `/tmp/testxml-20211212-10682-nobigf/testxml-0.0.1/target`
```

Jc

**Comment by danielnachun** [comment, marked as answer]
Two things here that might help:

1) You probably need to add `pkg-config` as a Linux only dependency.  That is how most Rust packages find `libxml2`.

2) You should check if the final binaries you build are dynamically linked to `libxml2`, or if it's just using the headers.  You can check what is dynamically linked to your binaries with `ldd`.  If `libxml2` is in that list, then `libxml2` needs to be a run time dependency as well on Linux, and you should remove `=> :build` after it.  Note that because `libxml2` is provided by macOS but not Linux, we typically use `uses_from_macos "libxml2"` (possibly followed by `=> :build`) instead, unless the version of `libxml2` provided by macOS is too old.

**Reply by jcamiel** [reply]
Hi @danielnachun ,

Thank you very much that's exactly what was missing! I've tested your suggestions and my formula is building on Linux now.
Just for reference, my minimal formula for a Rust binary with libxml crate on Linux:

```ruby
class Testxml < Formula
  desc "Test xml"
  url "https://github.com/jcamiel/testxml/archive/refs/tags/0.0.1.tar.gz"
  sha256 "72c145617d9af341d332c330f164a987a11b6abe590522a8b675ccd8f00edb78"

  depends_on "rust" => :build

  on_linux do
      depends_on "pkg-config" => :build
      depends_on "libxml2"
  end

  def install
      system "cargo", "install", *std_cargo_args(path: ".")
  end

end
```

And the final binary is dynamically linked against `libxml2`:

```
$ ldd /home/linuxbrew/.linuxbrew/bin/testxml
	linux-vdso.so.1 (0x00007ffdccdba000)
	libxml2.so.2 => /home/linuxbrew/.linuxbrew/lib/libxml2.so.2 (0x00007f4d7be02000)
	libgcc_s.so.1 => /home/linuxbrew/.linuxbrew/lib/gcc/5/libgcc_s.so.1 (0x00007f4d7bde8000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4d7bdc2000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4d7bc7e000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f4d7bc78000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4d7bab1000)
	/home/linuxbrew/.linuxbrew/lib/ld.so => /lib64/ld-linux-x86-64.so.2 (0x00007f4d7c1c4000)
	libz.so.1 => /home/linuxbrew/.linuxbrew/lib/libz.so.1 (0x00007f4d7b89a000)
```

Thank you!!
